Fix crash when extending taken-over named class

KANAjetzt opened the pull request:
Added error handling in `_prepare_compilation()` to address cases where the `base_type` cannot be found, preventing a crash. 
Added an additional if statement to `find_class()` to check the script path.

I expect this to not be the ideal solution, as I stated in my issue comment.
But this PR might be a better place to discuss code changes, and I'm looking forward to any feedback.

Nonetheless, I ran all the GDScript tests and tried this fix in the MRP and two bigger projects with success and no obvious side effects.

Cheers.

---
> [!NOTE]
> This now only contains the crash prevention. Future debugging is needed. #83542 will stay open.

adamscott commented:
If we merge this PR, I'm wondering if we should update the GDScriptCache when a path is taken over by a new script.

adamscott reviewed (approved):
Seems fine!

Thank you again for your first PR!

akien-mga commented on modules/gdscript/gdscript_compiler.cpp:
```diff
@@ -2647,7 +2647,10 @@ Error GDScriptCompiler::_prepare_compilation(GDScript *p_script, const GDScriptP
 
 	GDScriptDataType base_type = _gdtype_from_datatype(p_class->base_type, p_script, false);
 
+	ERR_FAIL_COND_V_MSG(base_type.native_type == StringName(), ERR_BUG, vformat(R"(Failed to get base class for "%s")", p_script->path));
```
Should this be an engine-level error (`ERR_BUG`), or a user-level error (`_set_error` + returning the most relevant error code)?

dalexeev replied:
I think it's fine, this should not happen as a result of user error, it's an analyzer/compiler bug.

dalexeev commented on modules/gdscript/gdscript.cpp:
```diff
@@ -1102,6 +1102,9 @@ GDScript *GDScript::find_class(const String &p_qualified_name) {
 		// Script path could have a class path separator("::") in it.
 		class_names = p_qualified_name.trim_prefix(get_root_script()->path).split("::");
 		result = get_root_script();
+	} else if (path.begins_with(get_root_script()->path)) {
+		class_names = path.trim_prefix(get_root_script()->path).split("::");
+		result = get_root_script();
```
I'm not sure, but the condition looks as always true to me. `p_qualified_name` is not used at all in the branch, this is weird.

KANAjetzt replied:
Thanks for looking at this! I tried to explain what I found while debugging [here](https://github.com/godotengine/godot/issues/83542#issuecomment-1782628013).

If you can provide more details, I can dive back in.

"_The condition looks as always true_" - Which condition?
_"`p_qualified_name` is not used_" - What exactly do you mean by "not used" and what is weird about it?

dalexeev replied:
> Which condition?

`path.begins_with(get_root_script()->path)`. `path` is the script path, for inner classes it may additionally contain suffixes. `get_root_script()` returns the topmost script that owns the inner class scripts.

> What exactly do you mean by "not used" and what is weird about it?

This method checks `p_qualified_name`, every other branch has `p_qualified_name`, but not this one, that's weird. I haven't tested this, but it might break the `_owner != nullptr` branch.

KANAjetzt replied:
>  I haven't tested this, but it might break the _owner != nullptr branch.

How can I test for this? 
Adding some inner classes to my MRP?
I run `--test --test-suite="*GDScript*"` and everything seems to be fine.
![image](https://github.com/godotengine/godot/assets/41547570/57d3ef55-6eca-4100-8c68-234ba695c1f2)

dalexeev replied:
Successful tests do not guarantee the absence of regressions, since the tests do not cover all possible scenarios. I added this PR for discussion at the GDScript team meeting or maybe @vnen will review it offline.

vnen replied:
It is odd to me as well. This is a function to find a class identified by `p_qualified_name`, so the branch not checking the argument cannot reliably find the requested class. As @dalexeev said, this will always be true (the path of an inner class will always start with the path of its root script), so the other conditions will never be checked.

Given that, I don't think this is a proper solution.

vnen reviewed (dismissed):
See comment

pirey0 commented:
@vnen i've been talking with Kana and further analyzed the root cause for this issue (see https://github.com/godotengine/godot/issues/83542#issuecomment-2094835838)
 This problem will not be fixed by changes to find_class(.
Maybe we should close this PR, and address the actual problem in a new PR?

KANAjetzt committed:
Fix crash when extending taken-over named class

Added error handling in `_prepare_compilation()` to address cases where the `base_type` cannot be found, preventing a crash.

akien-mga commented:
This seems good as crash prevention, but the bug still exsits. Do we have an open bug report with a MRP keeping track of it?

pirey0 commented:
> This seems good as crash prevention, but the bug still exsits. Do we have an open bug report with a MRP keeping track of it?

https://github.com/godotengine/godot/issues/83542

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:
